torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1117,6 +1119,153 @@ def arg_extractor(combine_fn, input, dim):
+                    "requires_grad": SourcelessBuilder.create(tx, inp.requires_grad),
+                },
+            )
+            for inp, ini in itertools.chain(
+                zip(input.items, init.items), zip(input.items, init.items)
+            )
+        ]
+        (
+            (combine_result, combine_treespec),
+            combine_graph,
+            combine_lifted_freevars,
+        ) = speculate_subgraph(
+            tx,
+            combine_fn,
+            sub_args,
+            sub_kwargs={},
+            description="scan_combine",
+            source_target=self.value,
+            set_subgraph_inputs="flatten_manual",
+        )
+
+        if combine_lifted_freevars:
+            unimplemented(
+                f"Combine fn had unexpected freevars: {combine_lifted_freevars}"
+            )
+
+        if combine_result.python_type() != list:
+            unimplemented(
+                f"Expected combine_fn to return a list if tensor but got {combine_result.python_type()}",
+            )
+
+        input_proxy = input.as_proxy()
+        init_proxy = init.as_proxy()
+        combine_result_proxy = combine_result.as_proxy()
+        for result, inp_proxy, ini_proxy, s_args in zip(
+            combine_result_proxy, input_proxy, init_proxy, sub_args
+        ):
+            inp_meta = inp_proxy.node.meta["example_value"]
+            ini_meta = ini_proxy.node.meta["example_value"]
+            combine_result_meta = result.node.meta["example_value"]
+            s_args = s_args.as_proxy().node.meta["example_value"]
+            if combine_result_meta.device != inp_meta.device:
+                unimplemented(
+                    f"Expected combine_fn to return a tensor on device {inp_meta.device} but "
+                    + f"got {combine_result_meta.device}"
+                )
+            if combine_result_meta.dtype != inp_meta.dtype:
+                unimplemented(
+                    f"Expected combine_fn to return a tensor of {inp_meta.dtype} but "
+                    + f"got {combine_result_meta.dtype}"
+                )
+            if combine_result_meta.shape != s_args.shape:
+                unimplemented(
+                    f"Expected shape of input with {s_args.shape} to be the same as "
+                    + f"combine_result with {combine_result_meta.shape}"
+                )
+            if (
+                combine_result_meta.device != ini_meta.device
+                or combine_result_meta.dtype != ini_meta.dtype
+                or combine_result_meta.shape != ini_meta.shape
+            ):
+                unimplemented(
+                    f"Expected metadata of the combine_fn result {combine_result_meta} to be the same as "
+                    + f"the metadata of init with {inp_meta}"
+                )
+
+        combine_gm = torch.fx.GraphModule(dict(tx.output.nn_modules), combine_graph)
+        combine_fn_name = add_subgraph(tx, "scan_combine", combine_gm)
```

Comment:
De-dup `associatie_scan_combine_fn` and `scan_combine_fn`?

Reply:
Done